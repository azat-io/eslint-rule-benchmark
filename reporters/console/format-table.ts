import pc from 'picocolors'

import type { SingleRuleResult } from '../../runners/run-single-rule'

/**
 * Formats benchmark results as a console-friendly table.
 *
 * @param {SingleRuleResult} result - The benchmark result to format.
 * @returns {string} Formatted table string ready for console output.
 */
export let formatTable = (result: SingleRuleResult): string => {
  let { summary, rule } = result

  let formatMs = (ms: undefined | number): string =>
    ms === undefined || !Number.isFinite(ms) ? 'N/A' : `${ms.toFixed(2)} ms`

  let formatNumber = (number_: number): string => number_.toString()

  let columnWidth = 20
  let valueColumnWidth = 15

  let padColumn = (text: string): string => text.padEnd(columnWidth)

  let header = pc.bold(`Rule Benchmark Results: ${pc.cyan(rule.id)}`)

  let columnHeaders = `${pc.bold(padColumn('Metric'))} ${pc.bold('Value')}`
  let separator = pc.bold('='.repeat(columnWidth + valueColumnWidth))

  let rows = [
    `${padColumn('Mean time')} ${pc.green(formatMs(summary.meanTimeMs))}`,
    `${padColumn('Median time')} ${pc.green(formatMs(summary.medianTimeMs))}`,
    `${padColumn('Min time')} ${pc.green(formatMs(summary.minTimeMs))}`,
    `${padColumn('Max time')} ${pc.green(formatMs(summary.maxTimeMs))}`,
    `${padColumn('Total samples')} ${formatNumber(summary.totalSamples)}`,
    `${padColumn('Total warnings')} ${pc.yellow(formatNumber(summary.totalWarnings))}`,
    `${padColumn('Total errors')} ${pc.red(formatNumber(summary.totalErrors))}`,
  ]

  return [header, '', columnHeaders, separator, ...rows].join('\n')
}
