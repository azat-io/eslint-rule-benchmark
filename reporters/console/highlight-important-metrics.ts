import pc from 'picocolors'

import type { SingleRuleResult } from '../../runners/run-single-rule'

/** Options for highlighting important metrics. */
interface HighlightImportantMetricsOptions {
  /** Threshold for slow execution time in milliseconds. */
  slowThresholdMs?: number
}

/**
 * Highlights important metrics from a benchmark result.
 *
 * @param {SingleRuleResult} result - The benchmark result to analyze.
 * @param {HighlightImportantMetricsOptions} options - Options for highlighting.
 * @returns {string[]} Array of highlighted metrics as strings.
 */
export let highlightImportantMetrics = (
  result: SingleRuleResult,
  options: HighlightImportantMetricsOptions = {},
): string[] => {
  let { slowThresholdMs = 50 } = options
  let { summary } = result
  let highlights: string[] = []

  if (summary.meanTimeMs > slowThresholdMs) {
    let message = `⚠️  Mean execution time (${summary.meanTimeMs.toFixed(
      2,
    )} ms) exceeds threshold of ${slowThresholdMs} ms`

    highlights.push(pc.yellow(message))
  }

  if (summary.maxTimeMs > slowThresholdMs * 2) {
    let message = `⚠️  Max execution time (${summary.maxTimeMs.toFixed(
      2,
    )} ms) is significantly high`

    highlights.push(pc.yellow(message))
  }

  if (summary.totalErrors > 0) {
    let message = `ℹ️  Rule found ${summary.totalErrors} errors in test code`

    highlights.push(pc.cyan(message))
  }

  if (summary.totalWarnings > 0) {
    let message = `ℹ️  Rule found ${summary.totalWarnings} warnings in test code`

    highlights.push(pc.cyan(message))
  }

  if (highlights.length === 0) {
    let message = '✅ All metrics are within expected ranges'

    highlights.push(pc.green(message))
  }

  return highlights
}
