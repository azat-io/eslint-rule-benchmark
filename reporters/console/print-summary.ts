import pc from 'picocolors'

import type { SingleRuleResult } from '../../runners/run-single-rule'

import { highlightImportantMetrics } from './highlight-important-metrics'
import { formatTable } from './format-table'

/** Options for printing benchmark summary. */
interface PrintSummaryOptions {
  /** Threshold for slow execution time in milliseconds. */
  slowThresholdMs?: number
}

/**
 * Prints a summary of benchmark results to the console.
 *
 * @param {SingleRuleResult} result - The benchmark result to print.
 * @param {PrintSummaryOptions} options - Options for printing.
 * @returns {void}
 */
export let printSummary = (
  result: SingleRuleResult,
  options: PrintSummaryOptions = {},
): void => {
  let { slowThresholdMs = 50 } = options

  console.info(formatTable(result))

  console.info(`\n${pc.bold('Highlights:')}`)
  let highlights = highlightImportantMetrics(result, {
    slowThresholdMs,
  })

  for (let highlight of highlights) {
    console.info(highlight)
  }

  console.info('')
}
