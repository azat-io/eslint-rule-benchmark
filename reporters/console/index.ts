import type { SingleRuleResult } from '../../runners/run-single-rule'
import type { BenchmarkConfig } from '../../types/benchmark-config'

import { printSummary } from './print-summary'

/**
 * Creates a console reporter for benchmark results.
 *
 * @returns {(result: SingleRuleResult, config: BenchmarkConfig) => void} A
 *   function that reports benchmark results to the console.
 */
export let createConsoleReporter =
  (): ((result: SingleRuleResult, config: BenchmarkConfig) => void) =>
  (result: SingleRuleResult, config: BenchmarkConfig): void => {
    printSummary(result, {
      slowThresholdMs: config.timeout / 100,
    })
  }

/**
 * Console reporter factory function.
 *
 * @returns {(result: SingleRuleResult, config: BenchmarkConfig) => void}
 *   Reporter function.
 */
export let consoleReporter = (): ((
  result: SingleRuleResult,
  config: BenchmarkConfig,
) => void) => createConsoleReporter()
