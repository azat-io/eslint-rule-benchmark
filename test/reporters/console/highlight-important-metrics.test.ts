import { beforeEach, afterEach, describe, expect, it, vi } from 'vitest'

import { highlightImportantMetrics } from '../../../reporters/console/highlight-important-metrics'

vi.mock('picocolors', () => ({
  default: {
    yellow: (text: string) => text,
    green: (text: string) => text,
    cyan: (text: string) => text,
  },
}))

describe('highlightImportantMetrics', () => {
  beforeEach(() => {
    vi.spyOn(console, 'info').mockImplementation(() => {})
  })

  afterEach(() => {
    vi.restoreAllMocks()
  })

  it('should highlight important metrics', () => {
    let result = highlightImportantMetrics({
      summary: {
        totalSamples: 1000,
        medianTimeMs: 200,
        totalWarnings: 10,
        meanTimeMs: 100,
        maxTimeMs: 300,
        totalErrors: 5,
        minTimeMs: 50,
      },
      rule: {
        id: 'rule-1',
      },
      benchmarkResults: [],
    })
    expect(result).toMatchSnapshot()
  })

  it('should highlight important metrics with custom slow threshold', () => {
    let result = highlightImportantMetrics(
      {
        summary: {
          totalSamples: 1000,
          medianTimeMs: 200,
          totalWarnings: 10,
          meanTimeMs: 100,
          maxTimeMs: 300,
          totalErrors: 5,
          minTimeMs: 50,
        },
        rule: {
          id: 'rule-1',
        },
        benchmarkResults: [],
      },
      { slowThresholdMs: 150 },
    )
    expect(result).toMatchSnapshot()
  })

  it('should highlight important metrics with no warnings or errors', () => {
    let result = highlightImportantMetrics({
      summary: {
        totalSamples: 1000,
        medianTimeMs: 200,
        totalWarnings: 0,
        meanTimeMs: 50,
        maxTimeMs: 100,
        totalErrors: 0,
        minTimeMs: 50,
      },
      rule: {
        id: 'rule-1',
      },
      benchmarkResults: [],
    })
    expect(result).toMatchSnapshot()
  })
})
