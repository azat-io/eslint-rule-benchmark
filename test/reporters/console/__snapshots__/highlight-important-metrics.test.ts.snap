// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`highlightImportantMetrics > should highlight important metrics 1`] = `
[
  "⚠️  Mean execution time (100.00 ms) exceeds threshold of 50 ms",
  "⚠️  Max execution time (300.00 ms) is significantly high",
  "ℹ️  Rule found 5 errors in test code",
  "ℹ️  Rule found 10 warnings in test code",
]
`;

exports[`highlightImportantMetrics > should highlight important metrics with custom slow threshold 1`] = `
[
  "ℹ️  Rule found 5 errors in test code",
  "ℹ️  Rule found 10 warnings in test code",
]
`;

exports[`highlightImportantMetrics > should highlight important metrics with no warnings or errors 1`] = `
[
  "✅ All metrics are within expected ranges",
]
`;
